File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/DefaultPaginationDescription.java
package com.azure.android.core.util.paging;

import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import java.util.concurrent.Executor;

/**
 * THE TYPE IS FOR INTERNAL USE.
 * (Consumed only by client-sdk but exposed to the user as {@link PaginationDescription} ).
 *
 * A default implementation of {@link PaginationDescription} returned from
 * {@link PaginationDescriptionRepository#get(Object)}.
 *
 * @param <T> the type of items in page
 */
public class DefaultPaginationDescription<T> implements PaginationDescription<T> {
    private final LiveData<PagedList<T>> pagedListObservable;
    private final LiveData<DataSourceNotification> dataSourceNotificationObservable;
    private final LiveData<PageLoadState> loadStateObservable;
    private final LiveData<PageLoadState> refreshStateObservable;
    private final Executor retryExecutor;

    private DefaultPaginationDescription(LiveData<DataSourceNotification> dataSourceNotificationObservable,
                              LiveData<PagedList<T>> pagedListObservable,
                              Executor retryExecutor) {
        this.dataSourceNotificationObservable = dataSourceNotificationObservable;
        this.pagedListObservable = pagedListObservable;
        this.retryExecutor = retryExecutor;
        this.loadStateObservable = DataSourceNotification.getLoadStateObservable(dataSourceNotificationObservable);
        this.refreshStateObservable = DataSourceNotification.getRefreshStateObservable(dataSourceNotificationObservable);
    }

    @Override
    public LiveData<PagedList<T>> getPagedListObservable() {
        return this.pagedListObservable;
    }

    @Override
    public LiveData<PageLoadState> getLoadStateObservable() {
        return this.loadStateObservable;
    }

    @Override
    public LiveData<PageLoadState> getRefreshStateObservable()  {
        return this.refreshStateObservable;
    }

    @Override
    public void retry() {
        DataSourceNotification notification = this.dataSourceNotificationObservable.getValue();
        if (notification != null) {
            Runnable retryRunnable = notification.getRetryRunnable();
            if (retryRunnable != null) {
                this.retryExecutor.execute(retryRunnable);
            }
        }
    }

    @Override
    public void refresh() {
        DataSourceNotification notification = this.dataSourceNotificationObservable.getValue();
        if (notification != null) {
            notification.invalidateDataSource();
        }
    }

    public static <T> PaginationDescription<T> create(PageItemsFetcher<T> pageItemsFetcher, PaginationOptions options) {
        PagedList.Config config = options.getPagedListConfig();
        if (config == null) {
            config = new PagedList.Config.Builder().build();
        }

        PageItemsFetcherBasedDataSourceFactory dataSourceFactory = new PageItemsFetcherBasedDataSourceFactory(pageItemsFetcher);

        LiveData<PagedList<T>> pagedListObservable = new LivePagedListBuilder(dataSourceFactory, config)
                    .setFetchExecutor(options.getPageLoadExecutor())
                    .build();

        LiveData<DataSourceNotification> dataSourceNotificationObservable = dataSourceFactory.getDataSourceNotificationObservable();
        return new DefaultPaginationDescription<>(dataSourceNotificationObservable,
                pagedListObservable,
                options.getPageLoadExecutor());
    }
}

File: samples/sample-app-storage-kotlin/src/main/java/com/azure/android/storage/sample/kotlin/core/util/paging/DefaultPaginationDescription.kt
package com.azure.android.storage.sample.kotlin.core.util.paging

import androidx.lifecycle.LiveData
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import java.util.concurrent.Executor

/**
 * THE TYPE IS FOR INTERNAL USE.
 * (Consumed only by client-sdk but exposed to the user as [PaginationDescription] ).
 *
 * A default implementation of [PaginationDescription] returned from
 * [PaginationDescriptionRepository.get].
 *
 * @param T the type of items in page
 */
class DefaultPaginationDescription<T> private constructor(mDataSourceNotificationObservable: LiveData<DataSourceNotification>,
                                                          mPagedListObservable: LiveData<PagedList<T>>,
                                                          mRetryExecutor: Executor) : PaginationDescription<T> {

    private val mDataSourceNotificationObservable: LiveData<DataSourceNotification> = mDataSourceNotificationObservable
    private val mPagedListObservable: LiveData<PagedList<T>> = mPagedListObservable
    private val mRetryExecutor: Executor = mRetryExecutor

    private val mLoadStateObservable: LiveData<PageLoadState>
        = DataSourceNotification.getLoadStateObservable(mDataSourceNotificationObservable)
    private val mRefreshStateObservable: LiveData<PageLoadState>
        = DataSourceNotification.getRefreshStateObservable(mDataSourceNotificationObservable)

    override val pagedListObservable: LiveData<PagedList<T>>
        get() = mPagedListObservable

    override val loadStateObservable: LiveData<PageLoadState>
        get() = mLoadStateObservable

    override val refreshStateObservable: LiveData<PageLoadState>
        get() = mRefreshStateObservable

    override fun retry() {
        val notification: DataSourceNotification? = mDataSourceNotificationObservable.value
        if (notification != null) {
            val retryRunnable: Runnable? = notification.getRetryRunnable()
            if (retryRunnable != null) {
                mRetryExecutor.execute(retryRunnable)
            }
        }
    }

    override fun refresh() {
        val notification: DataSourceNotification? = mDataSourceNotificationObservable.value
        notification?.invalidateDataSource()
    }

    companion object {
        fun <T> create(pageItemsFetcher: PageItemsFetcher<T>, options: PaginationOptions): PaginationDescription<T> {
            var config: PagedList.Config? = options.pagedListConfig
            if (config == null) {
                config = PagedList.Config.Builder().build()
            }
            val dataSourceFactory = PageItemsFetcherBasedDataSourceFactory(pageItemsFetcher)
            val pagedListObservable: LiveData<PagedList<T>>
                = LivePagedListBuilder<String?, T>(dataSourceFactory, config)
                    .setFetchExecutor(options.pageLoadExecutor)
                    .build()
            val dataSourceNotificationObservable: LiveData<DataSourceNotification>
                = dataSourceFactory.getDataSourceNotificationObservable()
            return DefaultPaginationDescription(dataSourceNotificationObservable,
                pagedListObservable,
                options.pageLoadExecutor)
        }
    }
}
